libraries/AP_RangeFinder/AP_RangeFinder_Benewake.cpp

```diff
@@ -0,0 +1,165 @@
+                                                             benewake_model_type model) :
+    AP_RangeFinder_Backend(_state),
+    model_type(model)
+{
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance);
+    if (uart != nullptr) {
+        uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance));
+    }
+}
+
+/* 
+   detect if a Benewake rangefinder is connected. We'll detect by
+   trying to take a reading on Serial. If we get a result the sensor is
+   there.
+*/
+bool AP_RangeFinder_Benewake::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+// distance returned in reading_cm, signal_ok is set to true if sensor reports a strong signal
+bool AP_RangeFinder_Benewake::get_reading(uint16_t &reading_cm, bool &signal_ok)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    float sum_cm = 0;
+    uint16_t count = 0;
+    bool dist_reliable = false;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        char c = uart->read();
+        // if buffer is empty and this byte is 0x59, add to buffer
+        if (linebuf_len == 0) {
+            if (c == BENEWAKE_FRAME_HEADER) {
+                linebuf[linebuf_len++] = c;
+            }
+        } else if (linebuf_len == 1) {
+            // if buffer has 1 element and this byte is 0x59, add it to buffer
+            // if not clear the buffer
+            if (c == BENEWAKE_FRAME_HEADER) {
+                linebuf[linebuf_len++] = c;
+            } else {
+                linebuf_len = 0;
+                dist_reliable = false;
+            }
+        } else {
+            // add character to buffer
+            linebuf[linebuf_len++] = c;
+            // if buffer now has 9 items try to decode it
+            if (linebuf_len == BENEWAKE_FRAME_LENGTH) {
+                // calculate checksum
+                uint16_t checksum = 0;
+                for (uint8_t i=0; i<BENEWAKE_FRAME_LENGTH-1; i++) {
+                    checksum += linebuf[i];
+                }
+                // if checksum matches extract contents
+                if ((uint8_t)(checksum & 0xFF) == linebuf[BENEWAKE_FRAME_LENGTH-1]) {
+                    // tell caller we are receiving packets
+                    signal_ok = true;
+                    // calculate distance and add to sum
+                    uint16_t dist = ((uint16_t)linebuf[3] << 8) | linebuf[2];
+                    if (dist != 0xFFFF) {
+                        // TFmini has short distance mode (mm)
+                        if ((model_type == BENEWAKE_TFmini)) {
+                            if (linebuf[6] == 0x02) {
+                                dist *= 0.1f;
```

Comment:
In case of short range, 0 - 5 m
For long range, 1-12 m
I think that it is better to judge other values as abnormal values

Reply:
I hope that it will return a bad value if it's out-of-range but I'm not sure yet.  let's see how testing goes.